File: crates/algal/src/embeddings.rs
// Embedding providers, mirroring src/embeddings.ts: a provider-neutral
// interface with two backends — the Vercel AI Gateway `/embeddings`
// endpoint and the deterministic local hashed-trigram embedder (FNV-1a
// 64, L2 normalized, f32-rounded). The local backend produces
// bit-identical vectors to TypeScript, so derived indexes rank the same
// corpus identically in either runtime.

use crate::{Error, Result, contract::text};
use serde_json::{Value, json};
use std::time::Duration;

pub const GATEWAY_BASE: &str = "https://ai-gateway.vercel.sh/v1";
pub const GATEWAY_EMBED_MODEL: &str = "openai/text-embedding-3-small";
pub const LOCAL_DIM: usize = 384;
pub const LOCAL_MODEL: &str = "algal/local-trigram-384";

pub const MAX_TEXTS: usize = 256;
pub const MAX_TEXT_BYTES: usize = 65_536;
pub const MAX_DIMENSION: usize = 4096;
pub const MAX_RESPONSE_BYTES: usize = 8_388_608;
const LOCAL_MAX_TOKENS: usize = 16_384;

/// ASCII lowercase alnum tokenizer — identical to `embedTokens` in
/// src/embeddings.ts. `max` bounds tokens per text.
pub fn embed_tokens(text: &str, max: usize) -> Vec<String> {
    let lower = text.to_lowercase();
    let mut tokens = Vec::new();
    let mut start: Option<usize> = None;
    for (i, c) in lower.char_indices() {
        let alnum = c.is_ascii_lowercase() || c.is_ascii_digit();
        if alnum {
            if start.is_none() {
                start = Some(i);
            }
        } else if let Some(s) = start.take() {
            tokens.push(lower[s..i].to_owned());
            if tokens.len() >= max {
                return tokens;
            }
        }
    }
    if let Some(s) = start {
        tokens.push(lower[s..].to_owned());
    }
    tokens.truncate(max);
    tokens
}

/// FNV-1a 64-bit — the same wrap arithmetic as the TypeScript BigInt
/// implementation, so feature buckets agree bit-for-bit.
fn fnv1a(bytes: &[u8]) -> u64 {
    let mut h: u64 = 0xcbf29ce484222325;
    for b in bytes {
        h = (h ^ u64::from(*b)).wrapping_mul(0x100000001b3);
    }
    h
}

/// One hashed-trigram embedding, identical to `localVector`: character
/// trigrams over each token map to `±1` accumulators (`h % dim` position,
/// `h & 1` sign), L2 normalized and rounded through f32.
pub fn local_vector(text: &str, dim: usize) -> Result<Vec<f64>> {
    if dim == 0 || dim > MAX_DIMENSION {
        return Err(Error::invalid(format!("invalid embedding dimension {dim}")));
    }
    let mut vec = vec![0f64; dim];
    for token in embed_tokens(text, LOCAL_MAX_TOKENS) {
        let bytes = token.as_bytes();
        for i in 0..bytes.len().saturating_sub(2) {
            let h = fnv1a(&bytes[i..i + 3]);
            let idx = (h % dim as u64) as usize;
            vec[idx] += if h & 1 == 0 { 1.0 } else { -1.0 };
        }
    }
    let norm: f64 = vec.iter().map(|v| v * v).sum::<f64>().sqrt();
    Ok(vec
        .iter()
        .map(|v| {
            if norm == 0.0 {
                0.0
            } else {
                // f64 divide, then round to f32 — Math.fround(v / norm)
                f64::from((v / norm) as f32)
            }
        })
        .collect())
}

/// Cosine similarity — identical op order to the TypeScript `cosine`.
pub fn cosine(a: &[f64], b: &[f64]) -> f64 {
    let (mut dot, mut na, mut nb) = (0f64, 0f64, 0f64);
    for (x, y) in a.iter().zip(b.iter()) {
        dot += x * y;
        na += x * x;
        nb += y * y;
    }
    let denom = na.sqrt() * nb.sqrt();
    if denom == 0.0 { 0.0 } else { dot / denom }
}

/// Deterministic token overlap `|q ∩ d| / |q|` — the lexical half of the
/// hybrid score, identical to the TypeScript `tokenOverlap`.
pub fn token_overlap(query: &str, text: &str) -> f64 {
    let q: std::collections::BTreeSet<String> =
        embed_tokens(query, LOCAL_MAX_TOKENS).into_iter().collect();
    if q.is_empty() {
        return 0.0;
    }
    let d: std::collections::BTreeSet<String> =
        embed_tokens(text, LOCAL_MAX_TOKENS).into_iter().collect();
    let hit = q.iter().filter(|t| d.contains(*t)).count();
    hit as f64 / q.len() as f64
}

fn check_texts(texts: &[String]) -> Result<()> {
    if texts.is_empty() || texts.len() > MAX_TEXTS {
        return Err(Error::invalid(format!(
            "embed texts requires 1..{MAX_TEXTS}"
        )));
    }
    for t in texts {
        if t.is_empty() {
            return Err(Error::invalid("embed text must be nonempty"));
        }
        if t.len() > MAX_TEXT_BYTES {
            return Err(Error::limit("embed text bytes"));
        }
    }
    Ok(())
}

/// OpenAI-compatible gateway embeddings — `POST {base}/embeddings` with
/// `{model, input}` → `{data:[{index, embedding}]}` sorted by index.
/// Bounded, no redirects, error bodies withheld.
pub async fn gateway_embed(
    base_url: &str,
    model: &str,
    credential: &str,
    texts: &[String],
    deadline_ms: u64,
) -> Result<Vec<Vec<f64>>> {
    check_texts(texts)?;
    text(&json!(model), 128).map_err(|_| Error::invalid("embedding model"))?;
    if credential.is_empty() || credential.len() > 8192 || credential.contains(['\r', '\n']) {
        return Err(Error::invalid("gateway credential configuration"));
    }
    let url = format!("{}/embeddings", base_url.trim_end_matches('/'));
    let body = json!({"model":model,"input":texts});
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .timeout(Duration::from_millis(deadline_ms.max(1)))
        .build()
        .map_err(|_| Error::new("EFFECT_FAILED", "HTTP client initialization failed"))?;
    let mut response = client
        .post(&url)
        .json(&body)
        .bearer_auth(credential)
        .send()
        .await
        .map_err(|_| {
            Error::new("EFFECT_FAILED", "embedding request failed or timed out").uncertain()
        })?;
    if !response.status().is_success() {
        return Err(Error::new(
            "EFFECT_FAILED",
            format!(
                "embedding endpoint returned HTTP {}; redirects and error bodies are not accepted",
                response.status().as_u16()
            ),
        ));
    }
    if response
        .content_length()
        .is_some_and(|n| n > MAX_RESPONSE_BYTES as u64)
    {
        return Err(Error::limit("embedding response bytes").uncertain());
    }
    let mut bytes = Vec::new();
    while let Some(chunk) = response
        .chunk()
        .await
        .map_err(|_| Error::new("EFFECT_FAILED", "embedding body read failed").uncertain())?
    {
        if bytes.len().saturating_add(chunk.len()) > MAX_RESPONSE_BYTES {
            return Err(Error::limit("embedding response bytes").uncertain());
        }
        bytes.extend_from_slice(&chunk);
    }
    let raw: Value = serde_json::from_slice(&bytes)
        .map_err(|_| Error::new("EFFECT_UNPARSEABLE", "embedding response is not JSON"))?;
    let data = raw["data"]
        .as_array()
        .ok_or_else(|| Error::new("EFFECT_UNPARSEABLE", "embedding data missing"))?;
    if data.len() != texts.len() {
        return Err(Error::new(
            "EFFECT_UNPARSEABLE",
            format!(
                "embedding response returned {} vectors for {} texts",
                data.len(),
                texts.len()
            ),
        ));
    }
    let mut vectors: Vec<Option<Vec<f64>>> = vec![None; data.len()];
    let mut dimension: Option<usize> = None;
    for entry in data {
        let index = entry["index"]
            .as_u64()
            .filter(|i| (*i as usize) < data.len())
            .ok_or_else(|| Error::new("EFFECT_UNPARSEABLE", "embedding index invalid"))?
            as usize;
        let vec: Vec<f64> = entry["embedding"]
            .as_array()
            .ok_or_else(|| Error::new("EFFECT_UNPARSEABLE", "embedding vector missing"))?
            .iter()
            .map(|v| {
                v.as_f64()
                    .filter(|n| n.is_finite())
                    .ok_or_else(|| Error::new("EFFECT_UNPARSEABLE", "embedding value invalid"))
            })
            .collect::<Result<_>>()?;
        if vec.is_empty() || vec.len() > MAX_DIMENSION {
            return Err(Error::new(
                "EFFECT_UNPARSEABLE",
                "embedding dimension invalid",
            ));
        }
        match dimension {
            None => dimension = Some(vec.len()),
            Some(d) if d == vec.len() => (),
            _ => {
                return Err(Error::new(
                    "EFFECT_UNPARSEABLE",
                    "embedding dimensions differ",
                ));
            }
        }
        vectors[index] = Some(vec);
    }
    vectors
        .into_iter()
        .collect::<Option<Vec<_>>>()
        .ok_or_else(|| Error::new("EFFECT_UNPARSEABLE", "embedding response gaps"))
}

/// The resolved embedder for a CLI invocation: `local` (default) or
/// `gateway[:<model>]`.
pub enum Embedder {
    Local { dim: usize },
    Gateway { model: String },
}

impl Embedder {
    pub fn resolve(spec: Option<&str>) -> Result<Self> {
        match spec {
            None | Some("local") => Ok(Self::Local { dim: LOCAL_DIM }),
            Some("gateway") => Ok(Self::Gateway {
                model: GATEWAY_EMBED_MODEL.into(),
            }),
            Some(spec) if spec.starts_with("gateway:") && spec.len() > "gateway:".len() => {
                Ok(Self::Gateway {
                    model: spec["gateway:".len()..].to_owned(),
                })
            }
            Some(spec) => Err(Error::invalid(format!(
                "unknown embedder \"{spec}\" (want local, gateway, or gateway:<model>)"
            ))),
        }
    }

    pub fn model(&self) -> String {
        match self {
            Self::Local { .. } => LOCAL_MODEL.into(),
            Self::Gateway { model } => model.clone(),
        }
    }

    pub fn dimension(&self) -> usize {
        match self {
            Self::Local { dim } => *dim,
            // gateway dimension is learned per response; report the
            // conventional size for status output
            Self::Gateway { .. } => 0,
        }
    }

    pub async fn embed(&self, texts: &[String], deadline_ms: u64) -> Result<Vec<Vec<f64>>> {
        match self {
            Self::Local { dim } => {
                check_texts(texts)?;
                texts.iter().map(|t| local_vector(t, *dim)).collect()
            }
            Self::Gateway { model } => {
                let credential = std::env::var("AI_GATEWAY_API_KEY")
                    .or_else(|_| std::env::var("VERCEL_OIDC_TOKEN"))
                    .map_err(|_| {
                        Error::new(
                            "EFFECT_UNBOUND",
                            "gateway credential is not configured — set AI_GATEWAY_API_KEY or VERCEL_OIDC_TOKEN",
                        )
                    })?;
                gateway_embed(GATEWAY_BASE, model, &credential, texts, deadline_ms).await
            }
        }
    }
}

File: src/embeddings.ts
// Embedding providers — a provider-neutral interface with two
// implementations: the Vercel AI Gateway `/embeddings` endpoint and a
// deterministic local fallback (hashed trigram features, FNV-1a 64, L2
// normalized). The local provider needs no network and no credential; the
// Rust mirror produces bit-identical vectors, so indexes and ranks are
// cross-runtime reproducible.
//
// Embeddings are a derived artifact — they rank the semantic index
// (src/semantic.ts) and never enter manifests, digests, or receipts.

import { AlgalError } from "./errors";
import { boundedBytes } from "./io-runtime";
import { asArray, asObject, reqField } from "./values";

export const VERCEL_GATEWAY_BASE = "https://ai-gateway.vercel.sh/v1" as const;
export const GATEWAY_EMBED_MODEL = "openai/text-embedding-3-small" as const;

export const EMBED_BOUNDS = {
  maxTexts: 256,
  maxTextBytes: 65_536,
  maxDimension: 4096,
  maxResponseBytes: 8_388_608,
} as const;

/** Anything that maps bounded text to fixed-dimension vectors: a remote
 * endpoint (gateway) or the deterministic local embedder. */
export interface Embedder {
  /** Stable provider id recorded in the index's meta table. */
  id: string;
  /** Model/provider identity — indexes are scoped to it. */
  model: string;
  /** Vector dimensionality; fixed per provider. */
  dimension: number;
  embed(texts: string[], signal?: AbortSignal): Promise<number[][]>;
}

export function checkTexts(texts: string[], at = "embed texts"): string[] {
  if (texts.length === 0 || texts.length > EMBED_BOUNDS.maxTexts) {
    throw new AlgalError(
      "PARSE_FAILED",
      `${at} requires 1..${EMBED_BOUNDS.maxTexts} texts`,
    );
  }
  for (const [i, text] of texts.entries()) {
    if (typeof text !== "string" || text.length === 0) {
      throw new AlgalError("PARSE_FAILED", `${at}[${i}] must be nonempty text`);
    }
    if (Buffer.byteLength(text, "utf8") > EMBED_BOUNDS.maxTextBytes) {
      throw new AlgalError(
        "BUDGET_EXHAUSTED",
        `${at}[${i}] exceeds ${EMBED_BOUNDS.maxTextBytes} bytes`,
      );
    }
  }
  return texts;
}

// ------------------------------------------------------------ local hash ---

/** The local vector dimension — shared with the Rust mirror, never changed
 * without a new provider id. */
export const LOCAL_DIM = 384;
export const LOCAL_MODEL = "algal/local-trigram-384" as const;
const LOCAL_MAX_TOKENS = 16_384;

/** ASCII lowercase alnum tokenizer — identical in both runtimes. */
export function embedTokens(text: string, maxTokens = LOCAL_MAX_TOKENS): string[] {
  const lower = text.toLowerCase();
  const tokens: string[] = [];
  let start = -1;
  const flush = (end: number) => {
    if (start >= 0) tokens.push(lower.slice(start, end));
    start = -1;
  };
  for (let i = 0; i < lower.length; i++) {
    const c = lower.charCodeAt(i);
    const alnum =
      (c >= 97 && c <= 122) || (c >= 48 && c <= 57);
    if (alnum) {
      if (start < 0) start = i;
    } else if (start >= 0) {
      flush(i);
      if (tokens.length >= maxTokens) return tokens;
    }
  }
  flush(lower.length);
  return tokens.slice(0, maxTokens);
}

/** FNV-1a 64-bit — wraps through BigInt so TS and Rust agree bit-for-bit. */
function fnv1a(bytes: string): bigint {
  let h = 0xcbf29ce484222325n;
  const prime = 0x100000001b3n;
  const mask = 0xffffffffffffffffn;
  for (let i = 0; i < bytes.length; i++) {
    h = ((h ^ BigInt(bytes.charCodeAt(i) & 0xff)) * prime) & mask;
  }
  return h;
}

/** One hashed-trigram embedding: character trigrams over each token map to
 * `±1` feature accumulators (`h % dim` position, `h & 1` sign), then L2
 * normalize. Deterministic, symmetric, cross-runtime bit-identical. */
export function localVector(text: string, dimension = LOCAL_DIM): number[] {
  if (dimension <= 0 || dimension > EMBED_BOUNDS.maxDimension) {
    throw new AlgalError("PARSE_FAILED", `invalid embedding dimension ${dimension}`);
  }
  const vec = new Float64Array(dimension);
  const dim = BigInt(dimension);
  for (const token of embedTokens(text)) {
    for (let i = 0; i + 3 <= token.length; i++) {
      const h = fnv1a(token.slice(i, i + 3));
      const idx = Number(h % dim);
      vec[idx] = vec[idx]! + ((h & 1n) === 0n ? 1 : -1);
    }
  }
  let norm = 0;
  for (const v of vec) norm += v * v;
  norm = Math.sqrt(norm);
  const out = new Array<number>(dimension);
  for (let i = 0; i < dimension; i++) {
    // round through f32 — the same value both runtimes persist
    out[i] = norm === 0 ? 0 : Math.fround(vec[i]! / norm);
  }
  return out;
}

/** The deterministic local embedder — no network, no credential, identical
 * vectors to the Rust `local_embed`. Default whenever the gateway is not
 * configured. */
export function localEmbedder(dimension = LOCAL_DIM): Embedder {
  return {
    id: "local",
    model: LOCAL_MODEL,
    dimension,
    async embed(texts) {
      const checked = checkTexts(texts);
      return checked.map((text) => localVector(text, dimension));
    },
  };
}

// ------------------------------------------------------------ gateway -----

export type GatewayEmbedderOptions = {
  /** Gateway bearer — value or async resolver (env/vercel OIDC). */
  credential: string | (() => Promise<string>);
  model?: string;
  baseUrl?: string;
  fetch?: typeof globalThis.fetch;
  /** Response ceiling; defaults to EMBED_BOUNDS.maxResponseBytes. */
  maxResponseBytes?: number;
};

async function boundedJson(
  response: Response,
  maxBytes: number,
  at: string,
): Promise<unknown> {
  if (response.status >= 300 && response.status < 400) {
    throw new AlgalError("EFFECT_FAILED", `${at} redirects are forbidden`);
  }
  const declared = response.headers.get("content-length");
  if (declared !== null && Number(declared) > maxBytes) {
    throw new AlgalError("EFFECT_FAILED", `${at} exceeds ${maxBytes} bytes`);
  }
  const bytes = await boundedBytes(response.body, maxBytes, at);
  if (!response.ok) {
    throw new AlgalError(
      "EFFECT_FAILED",
      `${at} returned HTTP ${response.status}; response body withheld`,
    );
  }
  try {
    return JSON.parse(new TextDecoder().decode(bytes));
  } catch {
    throw new AlgalError("EFFECT_UNPARSEABLE", `${at} returned invalid JSON`);
  }
}

/** OpenAI-compatible gateway embeddings: `POST {base}/embeddings` with
 * `{model, input:[...]}` → `{data:[{index, embedding:[…]}]}` sorted by
 * index. Credentials resolve per call, never recorded. */
export function gatewayEmbedder(options: GatewayEmbedderOptions): Embedder {
  const model = options.model ?? GATEWAY_EMBED_MODEL;
  if (!/^[a-z0-9][a-z0-9._/-]{0,127}$/i.test(model)) {
    throw new AlgalError("PARSE_FAILED", `invalid embedding model "${model}"`);
  }
  const baseUrl = (options.baseUrl ?? VERCEL_GATEWAY_BASE).replace(/\/+$/, "");
  const fetcher = options.fetch ?? globalThis.fetch;
  const maxBytes = options.maxResponseBytes ?? EMBED_BOUNDS.maxResponseBytes;
  const key = async (): Promise<string> => {
    const value =
      typeof options.credential === "function"
        ? await options.credential()
        : options.credential;
    if (
      typeof value !== "string" ||
      value.length < 8 ||
      value.length > 8192 ||
      /[\r\n]/.test(value)
    ) {
      throw new AlgalError(
        "EFFECT_UNBOUND",
        "gateway credential is not configured — set AI_GATEWAY_API_KEY or VERCEL_OIDC_TOKEN",
      );
    }
    return value;
  };
  return {
    id: "gateway",
    model,
    // dimension is learned from the first response; 0 until then
    dimension: 0,
    async embed(texts, signal) {
      const checked = checkTexts(texts);
      const response = await fetcher(`${baseUrl}/embeddings`, {
        method: "POST",
        redirect: "error",
        ...(signal ? { signal } : {}),
        headers: {
          authorization: `Bearer ${await key()}`,
          "content-type": "application/json",
        },
        body: JSON.stringify({ model, input: checked }),
      });
      const raw = asObject(await boundedJson(response, maxBytes, "embedding response"), "embedding response");
      const data = asArray(reqField(raw, "data", "embedding response"), "embedding data");
      if (data.length !== checked.length) {
        throw new AlgalError(
          "EFFECT_UNPARSEABLE",
          `embedding response returned ${data.length} vectors for ${checked.length} texts`,
        );
      }
      const vectors: (number[] | undefined)[] = new Array(data.length);
      let dimension: number | undefined;
      for (const entry of data) {
        const obj = asObject(entry, "embedding data[]");
        const index = obj.index;
        if (!Number.isSafeInteger(index) || (index as number) < 0 || (index as number) >= data.length) {
          throw new AlgalError("EFFECT_UNPARSEABLE", "embedding data index invalid");
        }
        const vec = asArray(reqField(obj, "embedding", "embedding data[]"), "embedding vector")
          .map((v, i) => {
            if (typeof v !== "number" || !Number.isFinite(v)) {
              throw new AlgalError("EFFECT_UNPARSEABLE", `embedding vector[${i}] invalid`);
            }
            return v;
          });
        if (vec.length === 0 || vec.length > EMBED_BOUNDS.maxDimension) {
          throw new AlgalError("EFFECT_UNPARSEABLE", "embedding dimension invalid");
        }
        if (dimension === undefined) dimension = vec.length;
        if (vec.length !== dimension) {
          throw new AlgalError("EFFECT_UNPARSEABLE", "embedding dimensions differ");
        }
        vectors[index as number] = vec;
      }
      this.dimension = dimension ?? 0;
      return vectors.map((v) => v!);
    },
  };
}

/** Gateway credential env chain — matches the gateway effect backend. */
export function gatewayCredential(): string | undefined {
  return process.env.AI_GATEWAY_API_KEY ?? process.env.VERCEL_OIDC_TOKEN;
}

/** Validate an embedder spec and return it normalized — `local`, `gateway`,
 * or `gateway:<model>`. Used at contract boundaries where the spec is data. */
export function checkEmbedderSpec(spec: string, at: string): string {
  if (spec === "local" || spec === "gateway") return spec;
  if (spec.startsWith("gateway:") && spec.length > "gateway:".length) {
    return spec;
  }
  throw new AlgalError(
    "PARSE_FAILED",
    `${at}: unknown embedder "${spec}" (want local, gateway, or gateway:<model>)`,
  );
}

/** Resolve the embedder for a CLI invocation: the gateway model when one
 * is named, the deterministic local embedder otherwise. */
export function resolveEmbedder(spec?: string): Embedder {
  const normalized = checkEmbedderSpec(spec ?? "local", "embedder");
  if (normalized === "local") return localEmbedder();
  if (normalized === "gateway") {
    return gatewayEmbedder({ credential: credentialResolver });
  }
  return gatewayEmbedder({
    credential: credentialResolver,
    model: normalized.slice("gateway:".length),
  });
}

async function credentialResolver(): Promise<string> {
  const value = gatewayCredential();
  if (value === undefined) {
    throw new AlgalError(
      "EFFECT_UNBOUND",
      "gateway credential is not configured — set AI_GATEWAY_API_KEY or VERCEL_OIDC_TOKEN",
    );
  }
  return value;
}

// ------------------------------------------------------------ ranking -----

export function cosine(a: number[], b: number[]): number {
  let dot = 0;
  let na = 0;
  let nb = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const x = a[i]!;
    const y = b[i]!;
    dot += x * y;
    na += x * x;
    nb += y * y;
  }
  const denom = Math.sqrt(na) * Math.sqrt(nb);
  return denom === 0 ? 0 : dot / denom;
}

/** Deterministic token overlap `|q ∩ d| / |q|` — the lexical half of the
 * hybrid score, computed identically in both runtimes. */
export function tokenOverlap(query: string, text: string): number {
  const q = new Set(embedTokens(query));
  if (q.size === 0) return 0;
  const d = new Set(embedTokens(text));
  let hit = 0;
  for (const t of q) if (d.has(t)) hit++;
  return hit / q.size;
}
